src/Microsoft.Health.Fhir.SqlServer/Features/Storage/SqlRetry/SqlRetry.cs

```diff
@@ -0,0 +1,280 @@
+
+                // Additional Fhir Server errors:
+                    8623,   // The query processor ran out of internal resources and could not produce a query plan.
+                };
+
+        private readonly IsExceptionRetriable defaultIsExceptionRetriable = DefaultIsExceptionRetriable;
+        private readonly bool defaultIsExceptionRetriableOff;
+        private readonly IsExceptionRetriable customIsExceptionRetriable;
+        private readonly int maxRetries;
+        private readonly int retryMillisecondsDelay;
+
+        public SqlRetryService(
+            ISqlConnectionBuilder sqlConnectionBuilder,
+            IOptions<SqlServerDataStoreConfiguration> sqlServerDataStoreConfiguration,
+            IOptions<SqlRetryServiceOptions> sqlRetryServiceOptions,
+            SqlRetryServiceDelegateOptions sqlRetryServiceDelegateOptions)
+        {
+            EnsureArg.IsNotNull(sqlConnectionBuilder, nameof(sqlConnectionBuilder));
+            _sqlServerDataStoreConfiguration = EnsureArg.IsNotNull(sqlServerDataStoreConfiguration?.Value, nameof(sqlServerDataStoreConfiguration));
+            _sqlConnectionBuilder = sqlConnectionBuilder;
+
+            EnsureArg.IsNotNull(sqlRetryServiceOptions?.Value, nameof(sqlRetryServiceOptions));
+            if (sqlRetryServiceOptions.Value.RemoveTransientErrors != null)
+            {
+                transientErrors.ExceptWith(sqlRetryServiceOptions.Value.RemoveTransientErrors);
+            }
+
+            if (sqlRetryServiceOptions.Value.AddTransientErrors != null)
+            {
+                transientErrors.UnionWith(sqlRetryServiceOptions.Value.AddTransientErrors);
+            }
+
+            maxRetries = sqlRetryServiceOptions.Value.MaxRetries;
+            retryMillisecondsDelay = sqlRetryServiceOptions.Value.RetryMillisecondsDelay;
+
+            EnsureArg.IsNotNull(sqlRetryServiceDelegateOptions, nameof(sqlRetryServiceDelegateOptions));
+            defaultIsExceptionRetriableOff = sqlRetryServiceDelegateOptions.DefaultIsExceptionRetriableOff;
+            customIsExceptionRetriable = sqlRetryServiceDelegateOptions.CustomIsExceptionRetriable;
+        }
+
+        public delegate bool IsExceptionRetriable(Exception ex);
+
+        public delegate Task ExecuteSqlWithRetriesAction(SqlCommand sqlCommand, CancellationToken cancellationToken);
+
+        public delegate Task<bool> ExecuteRowProcessor<T>(SqlDataReader reader, T result, CancellationToken cancellationToken);
+
+        public delegate T InitializeRowProcessor<T>();
+
+        public delegate void SqlConnectionInitializer(SqlConnection sqlConnection);
+
+        // TODO also define toT action delegate.
+
+        private static bool DefaultIsExceptionRetriable(Exception ex)
+        {
+            if (ex is SqlException sqlEx)
```

Comment:
>if (ex is SqlException sqlEx) [](http://example.com/codeflow?start=12&length=29)

Is this

Microsoft.Data.SqlClient.SqlException (0x80131904): A connection was successfully established with the server, but then an error occurred during the login process. (provider: TCP Provider, error: 0 - No error information)

retried?